src/utils/solanaHelpers.ts

```diff
@@ -319,35 +351,53 @@ export async function getEstimateBalanceChange(connection: Connection, tx: Trans
   }
 }
 
-export async function calculateTxFee(message: Message, connection: Connection): Promise<{ blockHash: string; fee: number; height: number }> {
+export async function calculateTxFee(
+  message: Message | VersionedMessage,
+  connection: Connection,
+  selectedAddress: string
+): Promise<{ blockHash: string; height: number; fee: number }> {
   const latestBlockHash = await connection.getLatestBlockhash("finalized");
   const blockHash = latestBlockHash.blockhash;
   const height = latestBlockHash.lastValidBlockHeight;
-  const fee = await connection.getFeeForMessage(message);
-  return { blockHash, fee: fee.value, height };
+
+  const instructions = decompile(message as unknown as VersionedMessage);
+  const legacyMessage = Message.compile({
+    instructions,
+    recentBlockhash: (message as unknown as VersionedMessage).recentBlockhash,
+    payerKey: new PublicKey(selectedAddress),
+  });
+  const fee = await connection.getFeeForMessage(legacyMessage);
+  return { blockHash, height, fee: fee.value || 0 };
 }
 
 export function decodeAllInstruction(messages: string[], messageOnly: boolean) {
   const decoded: DecodedDataType[] = [];
   (messages as string[]).forEach((msg) => {
-    let tx2: Transaction;
+    let tx2: VersionedTransaction;
     if (messageOnly) {
-      tx2 = Transaction.populate(Message.from(Buffer.from(msg, "hex")));
+      const msgObj = VersionedMessage.deserialize(Buffer.from(msg as string, "hex"));
+      tx2 = new VersionedTransaction(msgObj); // only for instuctions
     } else {
-      tx2 = Transaction.from(Buffer.from(msg, "hex"));
+      tx2 = VersionedTransaction.deserialize(Buffer.from(msg as string, "hex"));
     }
-    tx2.instructions.forEach((inst) => {
+    const instructions = decompile(tx2.message);
+    instructions.forEach((inst) => {
       decoded.push(decodeInstruction(inst));
     });
   });
   return decoded;
 }
 
-export function parsingTransferAmount(tx: Transaction, txFee: number, isGasless: boolean): FinalTxData | undefined {
-  if (tx.instructions.length > 1) return undefined;
-  if (!tx.instructions[0].programId.equals(SystemProgram.programId)) return undefined;
-  if (SystemInstruction.decodeInstructionType(tx.instructions[0]) !== "Transfer") return undefined;
-  const decoded = tx.instructions.map((inst) => {
+export function parsingTransferAmount(
+  tx: VersionedTransaction,
+  txFee: number,
+  isGasless: boolean,
+  instructions: TransactionInstruction[]
```

Comment:
instruction can be decompile from TransactionMessage using message in VersionTransaction